1-Draft/RFC00XX-Native-Command-Error-Handling.md

```diff
@@ -0,0 +1,229 @@
+---
+RFC: RFC00XX
+Author: Jason Helmick
+Status: Draft
+Area: Core
+Comments Due: 10/31/2020
+---
+
+# Native Command Error Handling
+
+PowerShell scripts using native commands would benefit from being able to use error handling
+features like those used by cmdlets.
+
+## Motivation
+
+In PowerShell by default, script processing continues when non-terminating errors occur. This is a
+benefit when expecting non-terminating errors in normal execution such as non-responsive computers from
+a list. This default behavior is controlled with the preference variable
+`$ErrorActionPreference` default of `Continue`.
+
+During development and debugging, often customers prefer that script execution stops when a
+non-terminating error occurs. PowerShell currently supports customers with this ability by setting
+the preference variable in the script.
+
+```Powershell
+$ErrorActionPreference = 'Stop'
+```
+
+Native commands return an exit code to the calling application which will be zero for success or
+non-zero for failure. However, native commands currently do not participate in the PowerShell error
+stream. Customers working with native commands in their scripts will need to check the execution
+status after each call using a helper function similar to below:
+
+```Powershell
+if( ! $? )
+{
+    exit $lastexitcode;
+}
+ ```
+
+Simply relaying the errors through the error stream isn't the solution. The example itself doesn't
+support all cases as `$?` can be false from a cmdlet or function error, making `$LASTEXITCODE`
+stale.
+
+To support a similar style of error handling with native commands, this specification proposes
+converting native command errors into PowerShell errors similar to the `sh`-compatible
+`set -e` mode.
+
+ The specification and alternative proposals are based on the
+ [Equivalent of bash `set -e` #3415](https://github.com/PowerShell/PowerShell/issues/3415)
+ committee review of the associated
+ [pull request](https://github.com/PowerShell/PowerShell/pull/3523), and
+ [implementation plan](https://github.com/PowerShell/PowerShell-RFC/pull/88#issuecomment-613653678)
+
+## Specification
+
+This RFC proposes including native commands in the error handling framework when the feature is
```

Comment:
I think `error handling framework` is somewhat amorphous language that won't help us crystalise a clear specification here.

Instead, maybe something like:

> This RFC proposes a preference variable to configure the elevation of errors produced by native commands to first-class PowerShell errors, so that native command failures will produce error objects that are added to the error stream and may terminate execution of the script without added boilerplate.